File: rust/src/problem/level_order.rs
use std::cell::RefCell;
use std::collections::VecDeque;
use std::rc::Rc;

use crate::problem::is_same_tree::TreeNode;

pub fn level_order(root: Option<Rc<RefCell<TreeNode>>>) -> Vec<Vec<i32>> {
    if root.is_none() {
        return vec![];
    }

    let mut result = vec![];
    let mut queue = VecDeque::from([root]);

    while !queue.is_empty() {
        let mut len = queue.len();
        let mut level = vec![];

        while len > 0 {
            if let Some(next) = queue.pop_front().unwrap() {
                level.push(next.borrow().val);

                if next.borrow().left.is_some() {
                    queue.push_back(next.borrow().left.clone())
                }

                if next.borrow().right.is_some() {
                    queue.push_back(next.borrow().right.clone())
                }
            }

            len -= 1;
        }

        result.push(level);
    }

    result
}

File: typescript/src/tree/levelOrder.ts
import type { TreeNode } from "./TreeNode.ts";

export function levelOrder(root: TreeNode | null): number[][] {
	if (!root) {
		return [];
	}

	const result: number[][] = [];
	const queue: TreeNode[] = [root];

	while (queue.length) {
		const level: number[] = [];

		for (let i = queue.length; i > 0; i--) {
			const node = queue.shift()!;
			level.push(node.val);

			node.left && queue.push(node.left);
			node.right && queue.push(node.right);
		}

		result.push(level);
	}

	return result;
}
